pkg/chain/ethereum/tbtc.go

```diff
@@ -318,49 +348,323 @@ func (tc *TbtcChain) SelectGroup(seed *big.Int) (chain.Addresses, error) {
+			)
+		}
+
+		signingMembersIndexes[i] = group.MemberIndex(memberIndex.Uint64())
+	}
+
+	return &tbtc.DKGChainResult{
+		SubmitterMemberIndex:     group.MemberIndex(result.SubmitterMemberIndex.Uint64()),
+		GroupPublicKey:           result.GroupPubKey,
+		MisbehavedMembersIndexes: result.MisbehavedMembersIndices,
+		Signatures:               result.Signatures,
+		SigningMembersIndexes:    signingMembersIndexes,
+		Members:                  result.Members,
+		MembersHash:              result.MembersHash,
+	}, nil
+}
+
+// convertTbtcDkgResultToEcdsaFormat converts the TBTC-specific DKG result to
+// the format applicable for the WalletRegistry.
+func convertDkgResultToChainFormat(
+	result *tbtc.DKGChainResult,
+) ecdsaabi.EcdsaDkgResult {
+	signingMembersIndices := make([]*big.Int, len(result.SigningMembersIndexes))
+	for i, memberIndex := range result.SigningMembersIndexes {
+		signingMembersIndices[i] = big.NewInt(int64(memberIndex))
+	}
+
+	return ecdsaabi.EcdsaDkgResult{
+		SubmitterMemberIndex:     big.NewInt(int64(result.SubmitterMemberIndex)),
+		GroupPubKey:              result.GroupPublicKey,
+		MisbehavedMembersIndices: result.MisbehavedMembersIndexes,
+		Signatures:               result.Signatures,
+		SigningMembersIndices:    signingMembersIndices,
+		Members:                  result.Members,
+		MembersHash:              result.MembersHash,
+	}
+}
+
+func validateMemberIndex(chainMemberIndex *big.Int) error {
+	maxMemberIndex := big.NewInt(group.MaxMemberIndex)
+	if chainMemberIndex.Cmp(maxMemberIndex) > 0 {
+		return fmt.Errorf("invalid member index value: [%v]", chainMemberIndex)
+	}
+
+	return nil
+}
+
+func (tc *TbtcChain) OnDKGResultChallenged(
+	handler func(event *tbtc.DKGResultChallengedEvent),
+) subscription.EventSubscription {
+	onEvent := func(
+		resultHash [32]byte,
+		challenger common.Address,
+		reason string,
+		blockNumber uint64,
+	) {
+		handler(&tbtc.DKGResultChallengedEvent{
+			ResultHash:  resultHash,
+			Challenger:  chain.Address(challenger.Hex()),
+			Reason:      reason,
+			BlockNumber: blockNumber,
+		})
+	}
+
+	return tc.walletRegistry.
+		DkgResultChallengedEvent(nil, nil, nil).
+		OnEvent(onEvent)
+}
+
+func (tc *TbtcChain) OnDKGResultApproved(
+	handler func(event *tbtc.DKGResultApprovedEvent),
+) subscription.EventSubscription {
+	onEvent := func(
+		resultHash [32]byte,
+		approver common.Address,
+		blockNumber uint64,
+	) {
+		handler(&tbtc.DKGResultApprovedEvent{
+			ResultHash:  resultHash,
+			Approver:    chain.Address(approver.Hex()),
+			BlockNumber: blockNumber,
+		})
+	}
+
+	return tc.walletRegistry.
+		DkgResultApprovedEvent(nil, nil, nil).
+		OnEvent(onEvent)
 }
 
-// TODO: Implement a real SubmitDKGResult action. The current implementation
-// just creates and pipes the DKG submission event to the handlers
-// registered in the dkgResultSubmissionHandlers map.
 func (tc *TbtcChain) SubmitDKGResult(
 	memberIndex group.MemberIndex,
-	result *dkg.Result,
+	tecdsaDkgResult *dkg.Result,
```

Comment:
I am super confused about this parameter and `convertDkgResultToChainFormat`. 

In `SubmitDKGResult` we take `dkg.Result` and convert it to `ecdsaabi.EcdsaDkgResult` to pass it to `walletRegistry.SubmitDkgResult`. At the same time, we have another function, `convertDkgResultToChainFormat` that accepts `*tbtc.DKGChainResult` and converts it to `ecdsaabi.EcdsaDkgResult`. That being said, I am confused about the relation of `tbtc.DKGChainResult` and `dkgResult`. If we expect all these functions to work with `tbtc.DKGChainResult`, why don't we expect the same from `SubmitDKGResult`?

```

	// IsDKGResultValid checks whether the submitted DKG result is valid from
	// the on-chain contract standpoint.
	IsDKGResultValid(dkgResult *DKGChainResult) (bool, error)

	// ChallengeDKGResult challenges the submitted DKG result.
	ChallengeDKGResult(dkgResult *DKGChainResult) error

	// ApproveDKGResult approves the submitted DKG result.
	ApproveDKGResult(dkgResult *DKGChainResult) error
```

Reply:
Maybe we should rename `tbtc.DKGChainResult` to `tbtc.SubmittedDKGChainResult`?

Reply:
Please also see [this](https://github.com/keep-network/keep-core/pull/3427#discussion_r1053172506) comment.

Reply:
I reworked it. See fdbd05f8 (also 07d4b585 and cb192599 to capture the big picture)